### PR TITLE
Phase2-hgx310A Extend HGCalWaferMask to provide list of corners of partial wafers with new definition of placement index

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Extend HGCalWaferMask to provide list of corners of partial wafers with new definition of placement index

#### PR validation:

Use the runTheMatrix test workflows 

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Nothing special

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cvuosalo @civanch @srimanob Please review this code and sign it in view of proceeding to the next step

File: Geometry/HGCalCommonData/src/HGCalWaferMask.cc [hunk 1]
```diff
@@ -6,7 +6,7 @@
 #include <algorithm>
 #include <sstream>
 
-//#define EDM_ML_DEBUG
+#define EDM_ML_DEBUG
```
Comment (civanch) [inline Geometry/HGCalCommonData/src/HGCalWaferMask.cc:9]: @bsunanda , you need to comment this out.

File: Geometry/HGCalCommonData/src/HGCalWaferMask.cc [hunk 2]
```diff
@@ -1151,3 +1151,246 @@ std::vector<std::pair<double, double> > HGCalWaferMask::waferXY(
...
+    int np[12][6] = {{5, 11, 8, 3, 4, 5},
+                     {0, 6, 9, 4, 5, 0},
+                     {1, 7, 10, 5, 0, 1},
+                     {2, 8, 11, 0, 1, 2},
+                     {3, 9, 6, 1, 2, 3},
+                     {4, 10, 7, 2, 3, 4},
+                     {1, 2, 3, 9, 6, 1},
+                     {0, 1, 2, 8, 11, 0},
+                     {5, 0, 1, 7, 10, 5},
+                     {6, 5, 0, 6, 9, 4},
```
Comment (rovere) [inline Geometry/HGCalCommonData/src/HGCalWaferMask.cc:1244]: There seems to be a genuine mistake in this definition.

**Review by rovere — CHANGES_REQUESTED** [review]

**Comment by rovere** [comment]
Thanks, @bsunanda for this PR. It would be nice, especially from the long term code maintenance point of view, to have some sort of documentation and explanation about the content of this PR. Slides, drawings or math explanations would equally do the job.

@cseez, @pfs

**Comment by bsunanda** [comment]
@cmsbuild Please Test

File: Geometry/HGCalCommonData/src/HGCalWaferMask.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalWaferMask.cc]: Corrected

Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalWaferMask.cc]: Done

File: Geometry/HGCalCommonData/src/HGCalWaferMask.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalWaferMask.cc]: Done

>

**Comment by bsunanda** [comment]
@cvuosalo @civanch @srimanob Please sign this PR so that we can proceed for the next step of V17 geometry

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by civanch** [comment]
+1

**Comment by rovere** [comment]
Ciao @bsunanda 
thanks for the prompt fix. It would still be useful to have some accompanying documentation, in the format you prefer, to complement this PR to explain what the "magic numbers" that appear mean.

Also, I see that you introduce an additional `waferXY` method, w/o modifying the existing one: is there a particular reason for this? Is the old method still needed somewhere? At some point some general cleaning of legacy/unused code is mandatory. 

For the _occasional reader_ this will generate quite some confusion to try and understand which one to use. Maybe adding a couple of lines as comments could help.

**Comment by bsunanda** [comment]
waferXY method gives corners of the partial wafers in its local coordinate system. We have moved from one set of conventions to a different one from V16 and V17. The two methods correspond to the 2 methods. Prithvi, Indra and I discussed the documentation and we shall provide it there. Once this PR is added to the master (which we need urgently so that Prithvi can add the algorithm and complete the wafer definition for v17), I shall add some inline documentation in the code.

**Comment by bsunanda** [comment]
@srimanob Please test

**Comment by bsunanda** [comment]
@srimanob Please sign this PR. We are waiting for adding the next steps

**Comment by bsunanda** [comment]
@srimanob Please sign this PR. We are waiting for adding the next steps

**Comment by srimanob** [comment]
+Upgrade

The PR introduce another way to get a list of corners of partial wafers. Documentation will be provided in the later PR.

**Comment by perrotta** [comment]
+1
- Please don't forget to add some inline documentation, as you are anticipating in https://github.com/cms-sw/cmssw/pull/37664#issuecomment-1111825280

**Pull request merged**